xarray/core/concat.py

```diff
@@ -263,100 +316,103 @@ def _dataset_concat(
     """
     from .dataset import Dataset
 
-    if compat not in ["equals", "identical"]:
-        raise ValueError(
-            "compat=%r invalid: must be 'equals' " "or 'identical'" % compat
-        )
-
     dim, coord = _calc_concat_dim_coord(dim)
     # Make sure we're working on a copy (we'll be loading variables)
     datasets = [ds.copy() for ds in datasets]
     datasets = align(
         *datasets, join=join, copy=False, exclude=[dim], fill_value=fill_value
     )
 
-    concat_over, equals = _calc_concat_over(datasets, dim, data_vars, coords)
+    result_dim_coords, concat_dim_lengths, dims_sizes, result_coord_names, data_names = _parse_datasets(
+        datasets, dim
+    )
+    dim_names = set(result_dim_coords)
+    unlabeled_dims = dim_names - result_coord_names
 
-    def insert_result_variable(k, v):
-        assert isinstance(v, Variable)
-        if k in datasets[0].coords:
-            result_coord_names.add(k)
-        result_vars[k] = v
+    both_data_and_coords = result_coord_names & data_names
+    if both_data_and_coords:
+        raise ValueError(
+            "%r is a coordinate in some datasets but not others."
+            % list(both_data_and_coords)[0]  # preserve format of error message
+        )
+    # we don't want the concat dimension in the result dataset yet
+    result_dim_coords.pop(dim, None)
+    dims_sizes.pop(dim, None)
+
+    # case where concat dimension is a coordinate but not a dimension
+    if dim in result_coord_names and dim not in dim_names:
+        datasets = [ds.expand_dims(dim) for ds in datasets]
+
+    # determine which variables to concatentate
+    concat_over, equals = _calc_concat_over(
+        datasets, dim, dim_names, data_vars, coords, compat
+    )
 
-    # create the new dataset and add constant variables
-    result_vars = OrderedDict()
-    result_coord_names = set(datasets[0].coords)
+    # determine which variables to merge, and then merge them according to compat
+    variables_to_merge = (result_coord_names | data_names) - concat_over - dim_names
+    if variables_to_merge:
+        to_merge = []
+        for ds in datasets:
+            if variables_to_merge - set(ds.variables):
+                raise ValueError(
+                    "encountered unexpected variables %r" % list(variables_to_merge)[0]
+                )
+            to_merge.append(ds.reset_coords()[list(variables_to_merge)])
+
+        merge_equals = {k: equals.get(k, None) for k in variables_to_merge}
+
+        result_vars = merge_variables(
+            expand_variable_dicts(to_merge),
+            priority_vars=None,
+            compat=compat,
+            equals=merge_equals,
```

Comment:
Rather than passing in this dict marking which variables are already known to be equal, it would be nice if we could avoid making these merge interface here more complex by adjusting the input dicts output by `expand_variable_dicts` instead.

Reply:
This doesn't work with `compat='noconflicts'` where `equals=True` but we need to use `fillna` to merge the variables. Supporting that basically requires that we reimplement `unique_variable` in `concat`.

Instead, I've changed this section to call `unique_variable` directly and pass it the `equals` argument, which seems cleaner than the current code. Does that look OK to you?

Reply:
The other point is that if we dont pass `equals` to `unique_variable` we run unnecessary computes. I think our options are to a) reimplement `unique_variable` in `concat` or b) pass `equals` to `unique_variable`.

Reply:
Yes, this looks good to me!